File: hanjo/개인용/programmers/구현/L2_다리를지나는트럭/Solution.java
package 구현.L2_다리를지나는트럭;

import java.util.*;

public class Solution {

    public static int solution(int bridge_length, int weight, int[] truck_weights) {

        Queue<Truck> waiting = new LinkedList<>();
        for (int truckWeight : truck_weights) {
            waiting.offer(new Truck(truckWeight, 0));
        }

        LinkedList<Truck> bridge = new LinkedList<>();
        bridge.offer(waiting.poll());
        int time = 1;

        while (!bridge.isEmpty()) {
            // 시간 증가 및 다리 무게 계산
            int weightSum = 0;
            for (Truck truck : bridge) {
                truck.increaseTime();
                weightSum += truck.weight;
            }
            time++;

            // 시간이 되면 다리에서 치움
            if (bridge.peek().isTimeOut(bridge_length)) {
                weightSum -= bridge.poll().weight;
            }

            // 무게가 된다면 트럭을 올림
            if (!waiting.isEmpty() && weightSum + waiting.peek().weight <= weight) {
                bridge.offer(waiting.poll());
            }
        }

        return time;
    }

    public static class Truck {
        public int weight;
        public int time;

        public Truck(int weight, int time) {
            this.weight = weight;
            this.time = time;
        }

        public void increaseTime() {
            time+=1;
        }

        public boolean isTimeOut(int bridgeLength) {
            return time == bridgeLength;
        }
    }

    public static void main(String[] args) {
        System.out.println(solution(2, 10, new int[]{7,4,5,6}));
        System.out.println(solution(100, 100, new int[]{10}));
        System.out.println(solution(100, 100, new int[]{10,10,10,10,10,10,10,10,10,10}));

    }
}

/**
 * ================================================================================
 * 링크 : https://programmers.co.kr/learn/courses/30/lessons/42583?language=java
 * 성공여부 : 성공
 * 풀이시간 : 1h
 * 
 * 시간복잡도 : O( bridge_length * 트럭개수 )
 * 메모리(테케5) : 94.9 MB
 * 소요 시간 : 38.56 ms
 * ================================================================================
 * 
 * 큐를 활용해야하는 문제
 * 객체지향적으로 품
 * for문을 최대한 적게쓰려고 노력함
 * 
 */

File: hanjo/개인용/programmers/구현/L2_다리를지나는트럭/solution.py
def solution(bridge_length, weight, truck_weights):
    l = len(truck_weights)
    time = 1
    passed_truck, now_on, now_on_time =[], [truck_weights.pop(0)], [0]
    while len(passed_truck) != l:
        # 시간 증가
        time +=1
        for i, n in enumerate(now_on_time):
            now_on_time[i] += 1
        # 시간지나서 통과한 트럭 확인
        if now_on_time:
            if now_on_time[0] == bridge_length:
                passed_truck.append(now_on.pop(0))
                now_on_time.pop(0)
            # 추가가능하다면 올림
            if truck_weights:
                if sum(now_on) + truck_weights[0] <= weight:
                    now_on.append(truck_weights.pop(0))
                    now_on_time.append(0)
    return time


print(solution(2, 10, [7,4,5,6]), 8)
print(solution(100, 100, [10]), 101)
print(solution(100, 100, [10,10,10,10,10,10,10,10,10,10]), 110)